Null course ID

java.lang.AssertionError: null course id
	at teammates.common.util.Assumption.fail(Assumption.java:59)
	at teammates.common.util.Assumption.assertTrue(Assumption.java:27)
	at teammates.common.util.Assumption.assertNotNull(Assumption.java:269)
	at teammates.ui.controller.InstructorFeedbackEditCopyAction.execute(InstructorFeedbackEditCopyAction.java:28)
	at teammates.ui.controller.Action.executeAndPostProcess(Action.java:337)
	at teammates.ui.controller.ControllerServlet.doPost(ControllerServlet.java:59)
	at teammates.ui.controller.ControllerServlet.doGet(ControllerServlet.java:38)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:617)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:717)
	at org.mortbay.jetty.servlet.ServletHolder.handle(ServletHolder.java:511)
	at org.mortbay.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1166)
	at teammates.storage.datastore.DatastoreFilter.doFilter(DatastoreFilter.java:28)
	at org.mortbay.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1157)
	at com.google.appengine.tools.appstats.AppstatsFilter.doFilter(AppstatsFilter.java:142)
	at org.mortbay.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1157)
	at com.google.apphosting.utils.servlet.ParseBlobUploadFilter.doFilter(ParseBlobUploadFilter.java:125)
	at org.mortbay.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1157)
	at com.google.apphosting.runtime.jetty.SaveSessionFilter.doFilter(SaveSessionFilter.java:35)
	at org.mortbay.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1157)
	at com.google.apphosting.utils.servlet.JdbcMySqlConnectionCleanupFilter.doFilter(JdbcMySqlConnectionCleanupFilter.java:60)
	at org.mortbay.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1157)
	at com.google.apphosting.utils.servlet.TransactionCleanupFilter.doFilter(TransactionCleanupFilter.java:43)
	at org.mortbay.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1157)
	at org.mortbay.jetty.servlet.ServletHandler.handle(ServletHandler.java:388)
	at org.mortbay.jetty.security.SecurityHandler.handle(SecurityHandler.java:216)
	at org.mortbay.jetty.servlet.SessionHandler.handle(SessionHandler.java:182)
	at org.mortbay.jetty.handler.ContextHandler.handle(ContextHandler.java:765)
	at org.mortbay.jetty.webapp.WebAppContext.handle(WebAppContext.java:418)
	at com.google.apphosting.runtime.jetty.AppVersionHandlerMap.handle(AppVersionHandlerMap.java:257)
	at org.mortbay.jetty.handler.HandlerWrapper.handle(HandlerWrapper.java:152)
	at org.mortbay.jetty.Server.handle(Server.java:326)
	at org.mortbay.jetty.HttpConnection.handleRequest(HttpConnection.java:542)
	at org.mortbay.jetty.HttpConnection$RequestHandler.headerComplete(HttpConnection.java:923)
	at com.google.apphosting.runtime.jetty.RpcRequestParser.parseAvailable(RpcRequestParser.java:76)
	at org.mortbay.jetty.HttpConnection.handle(HttpConnection.java:404)
	at com.google.apphosting.runtime.jetty.JettyServletEngineAdapter.serviceRequest(JettyServletEngineAdapter.java:146)
	at com.google.apphosting.runtime.JavaRuntime$RequestRunnable.run(JavaRuntime.java:442)
	at com.google.tracing.TraceContext$TraceContextRunnable.runInContext(TraceContext.java:437)
	at com.google.tracing.TraceContext$TraceContextRunnable$1.run(TraceContext.java:444)
	at com.google.tracing.CurrentContext.runInContext(CurrentContext.java:230)
	at com.google.tracing.TraceContext$AbstractTraceContextCallback.runInInheritedContextNoUnref(TraceContext.java:308)
	at com.google.tracing.TraceContext$AbstractTraceContextCallback.runInInheritedContext(TraceContext.java:300)
	at com.google.tracing.TraceContext$TraceContextRunnable.run(TraceContext.java:441)
	at com.google.apphosting.runtime.ThreadGroupPool$PoolEntry.run(ThreadGroupPool.java:234)
	at java.lang.Thread.run(Thread.java:745)

```
2015-09-01 08:45:28.846 /page/instructorFeedbackEditCopy 302 668ms 0kb Mozilla/5.0 (Windows NT 10.0; WOW64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/44.0.2403.157 Safari/537.36 module=default version=5-50
174.126.34.185 - hollyraye12 [31/Aug/2015:17:45:28 -0700] "GET /page/instructorFeedbackEditCopy HTTP/1.1" 302 117 - "Mozilla/5.0 (Windows NT 10.0; WOW64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/44.0.2403.157 Safari/537.36" "teammatesv4.appspot.com" ms=669 cpu_ms=314 cpm_usd=0.000492 app_engine_release=1.9.25 instance=00c61b117c36ae0e9e3931296243dddd8dae984b
I 2015-09-01 08:45:28.189
teammates.ui.controller.ControllerServlet doPost: Request received : https://teammatesv4.appspot.com/page/instructorFeedbackEditCopy:{}
I 2015-09-01 08:45:28.189
teammates.ui.controller.ControllerServlet doPost: User agent : Mozilla/5.0 (Windows NT 10.0; WOW64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/44.0.2403.157 Safari/537.36
I 2015-09-01 08:45:28.189
teammates.ui.controller.ActionFactory getAction: URL received :https://teammatesv4.appspot.com/page/instructorFeedbackEditCopy
I 2015-09-01 08:45:28.416
teammates.logic.core.Emails sendUsingGae: [Email sent]to=teammates@comp.nus.edu.sg|from=TEAMMATES Admin <Admin@teammatesv4.appspotmail.com>|subject=TEAMMATES (5.50): New System Exception: null course id
E 2015-09-01 08:45:28.809
teammates.logic.core.Emails sendErrorReport: Sent crash report: [Email sent]to=teammates@comp.nus.edu.sg|from=TEAMMATES Admin <Admin@teammatesv4.appspotmail.com>|subject=TEAMMATES (5.50): New System Exception: null course id
E 2015-09-01 08:45:28.809
teammates.ui.controller.ControllerServlet doPost: TEAMMATESLOG|||page|||System Error Report|||true|||Unknown|||Unknown|||Unknown|||Unknown|||<span class="text-danger">TEAMMATES (5.50): New System Exception: null course id</span><br><a href="#" onclick="showHideErrorMessage('error2117503');">Show/Hide Details >></a><br><span id="error2117503" style="display: none;"><b>Error Message</b> 
<br/> 
<pre><code> 
null course id 
</code></pre> 
<br/> 
 
<b>Request Path</b> 
/page 
<br/> 
 
<b>Request Parameters</b> 
{} 
<br/> 
 
<b>Stack Trace</b> 
<pre><code> 

java.lang.AssertionError: null course id
	at teammates.common.util.Assumption.fail(Assumption.java:59)
	at teammates.common.util.Assumption.assertTrue(Assumption.java:27)
	at teammates.common.util.Assumption.assertNotNull(Assumption.java:269)
	at teammates.ui.controller.InstructorFeedbackEditCopyAction.execute(InstructorFeedbackEditCopyAction.java:28)
	at teammates.ui.controller.Action.executeAndPostProcess(Action.java:337)
	at teammates.ui.controller.ControllerServlet.doPost(ControllerServlet.java:59)
	at teammates.ui.controller.ControllerServlet.doGet(ControllerServlet.java:38)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:617)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:717)
	at org.mortbay.jetty.servlet.ServletHolder.handle(ServletHolder.java:511)
	at org.mortbay.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1166)
	at teammates.storage.datastore.DatastoreFilter.doFilter(DatastoreFilter.java:28)
	at org.mortbay.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1157)
	at com.google.appengine.tools.appstats.AppstatsFilter.doFilter(AppstatsFilter.java:142)
	at org.mortbay.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1157)
	at com.google.apphosting.utils.servlet.ParseBlobUploadFilter.doFilter(ParseBlobUploadFilter.java:125)
	at org.mortbay.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1157)
	at com.google.apphosting.runtime.jetty.SaveSessionFilter.doFilter(SaveSessionFilter.java:35)
	at org.mortbay.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1157)
	at com.google.apphosting.utils.servlet.JdbcMySqlConnectionCleanupFilter.doFilter(JdbcMySqlConnectionCleanupFilter.java:60)
	at org.mortbay.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1157)
	at com.google.apphosting.utils.servlet.TransactionCleanupFilter.doFilter(TransactionCleanupFilter.java:43)
	at org.mortbay.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1157)
	at org.mortbay.jetty.servlet.ServletHandler.handle(ServletHandler.java:388)
	at org.mortbay.jetty.security.SecurityHandler.handle(SecurityHandler.java:216)
	at org.mortbay.jetty.servlet.SessionHandler.handle(SessionHandler.java:182)
	at org.mortbay.jetty.handler.ContextHandler.handle(ContextHandler.java:765)
	at org.mortbay.jetty.webapp.WebAppContext.handle(WebAppContext.java:418)
	at com.google.apphosting.runtime.jetty.AppVersionHandlerMap.handle(AppVersionHandlerMap.java:257)
	at org.mortbay.jetty.handler.HandlerWrapper.handle(HandlerWrapper.java:152)
	at org.mortbay.jetty.Server.handle(Server.java:326)
	at org.mortbay.jetty.HttpConnection.handleRequest(HttpConnection.java:542)
	at org.mortbay.jetty.HttpConnection$RequestHandler.headerComplete(HttpConnection.java:923)
	at com.google.apphosting.runtime.jetty.RpcRequestParser.parseAvailable(RpcRequestParser.java:76)
	at org.mortbay.jetty.HttpConnection.handle(HttpConnection.java:404)
	at com.google.apphosting.runtime.jetty.JettyServletEngineAdapter.serviceRequest(JettyServletEngineAdapter.java:146)
	at com.google.apphosting.runtime.JavaRuntime$RequestRunnable.run(JavaRuntime.java:468)
	at com.google.tracing.TraceContext$TraceContextRunnable.runInContext(TraceContext.java:437)
	at com.google.tracing.TraceContext$TraceContextRunnable$1.run(TraceContext.java:444)
	at com.google.tracing.CurrentContext.runInContext(CurrentContext.java:230)
	at com.google.tracing.TraceContext$AbstractTraceContextCallback.runInInheritedContextNoUnref(TraceContext.java:308)
	at com.google.tracing.TraceContext$AbstractTraceContextCallback.runInInheritedContext(TraceContext.java:300)
	at com.google.tracing.TraceContext$TraceContextRunnable.run(TraceContext.java:441)
	at com.google.apphosting.runtime.ThreadGroupPool$PoolEntry.run(ThreadGroupPool.java:234)
	at java.lang.Thread.run(Thread.java:745)
 
</code></pre> 
 
 
</span>|||/page/instructorFeedbackEditCopy
```

--- a/src/main/webapp/WEB-INF/web.xml
+++ b/src/main/webapp/WEB-INF/web.xml
@@ -309,7 +309,7 @@ http://java.sun.com/xml/ns/javaee/web-app_2_5.xsd"
           </user-data-constraint>
     </security-constraint>
     <error-page>
-        <error-code>403</error-code>
+        <error-code>401</error-code>
         <location>/unauthorized.jsp</location>
     </error-page>
     <error-page>

--- a/src/main/webapp/WEB-INF/web.xml
+++ b/src/main/webapp/WEB-INF/web.xml
@@ -313,6 +313,10 @@ http://java.sun.com/xml/ns/javaee/web-app_2_5.xsd"
         <location>/unauthorized.jsp</location>
     </error-page>
     <error-page>
+        <error-code>403</error-code>
+        <location>/unauthorized.jsp</location>
+    </error-page>
+    <error-page>
         <error-code>404</error-code>
         <location>/pageNotFound.jsp</location>
     </error-page>

--- a/src/main/webapp/deadlineExceededErrorPage.jsp
+++ b/src/main/webapp/deadlineExceededErrorPage.jsp
@@ -1,4 +1,5 @@
 <%@ page import="teammates.common.util.Const"%>
+<% response.setStatus(500);%>
 <!DOCTYPE html>
 
 <html>

--- a/src/main/webapp/entityNotFoundPage.jsp
+++ b/src/main/webapp/entityNotFoundPage.jsp
@@ -1,4 +1,5 @@
 <%@ page import="teammates.common.util.Const"%>
+<% response.setStatus(500);%>
 <!DOCTYPE html>
 <html>
 <head>

--- a/src/main/webapp/errorPage.jsp
+++ b/src/main/webapp/errorPage.jsp
@@ -1,4 +1,5 @@
 <%@ page import="teammates.common.util.Const"%>
+<% response.setStatus(500);%>
 <!DOCTYPE html>
 
 <html>

--- a/src/main/webapp/feedbackSessionNotVisible.jsp
+++ b/src/main/webapp/feedbackSessionNotVisible.jsp
@@ -1,5 +1,6 @@
 <%@ page import="teammates.common.util.Const"%>
 <%@ page import="teammates.common.datatransfer.FeedbackSessionAttributes"%>
+<% response.setStatus(403);%>
 <!DOCTYPE html>
 <%
     String startTimeString = (String)session.getAttribute(Const.ParamsNames.FEEDBACK_SESSION_NOT_VISIBLE);

--- a/src/main/webapp/js/instructor.js
+++ b/src/main/webapp/js/instructor.js
@@ -143,8 +143,8 @@ function setupFsCopyModal() {
                 $('#courseList').html("Loading possible destination courses. Please wait ...<br><img class='margin-center-horizontal' src='/images/ajax-loader.gif'/>");
             },
             error: function() {
-                $('#courseList').html('Error retrieving course list.' + 
-                    'Please close the dialog window and try again.');
+                $('#courseList').html("<p id='fs-copy-modal-error'>Error retrieving course list." + 
+                    "Please close the dialog window and try again.</p>");
             },
             success: function(data) {
                 $('#courseList').html(data);

--- a/src/main/webapp/pageNotFound.jsp
+++ b/src/main/webapp/pageNotFound.jsp
@@ -1,4 +1,5 @@
 <%@ page import="teammates.common.util.Const"%>
+<% response.setStatus(404);%>
 <!DOCTYPE html>
 
 <html>

--- a/src/main/webapp/unauthorized.jsp
+++ b/src/main/webapp/unauthorized.jsp
@@ -1,5 +1,5 @@
 <%@ page import="teammates.common.util.Const"%>
-<% response.setStatus(403);%>
+<% response.setStatus(401);%>
 <!DOCTYPE html>
 
 <html>

--- a/src/main/webapp/unauthorized.jsp
+++ b/src/main/webapp/unauthorized.jsp
@@ -1,4 +1,5 @@
 <%@ page import="teammates.common.util.Const"%>
+<% response.setStatus(403);%>
 <!DOCTYPE html>
 
 <html>

--- a/src/test/java/teammates/test/cases/ui/browsertests/InstructorHomePageUiTest.java
+++ b/src/test/java/teammates/test/cases/ui/browsertests/InstructorHomePageUiTest.java
@@ -391,6 +391,17 @@ public class InstructorHomePageUiTest extends BaseUiTestCase {
         homePage.verifyStatus(Const.StatusMessages.FEEDBACK_SESSION_COPIED);
         
         homePage.goToPreviousPage(InstructorHomePage.class);
+        
+        ______TS("Failure case: Ajax error");
+        
+        // Change action link so that ajax will fail
+        homePage.changeFsCopyButtonActionLink(courseId, feedbackSessionName, "/page/nonExistentPage?");
+        // Click copy
+        homePage.clickFsCopyButton(courseId, feedbackSessionName);
+        // Wait for modal to appear and show error.
+        homePage.waitForModalErrorToLoad();
+        
+        
     }
 
     public void testDeleteCourseAction() throws Exception{

--- a/src/test/java/teammates/test/pageobjects/InstructorHomePage.java
+++ b/src/test/java/teammates/test/pageobjects/InstructorHomePage.java
@@ -325,6 +325,16 @@ public class InstructorHomePage extends AppPage {
         return browser.driver.findElement(locator).getText();
     }
     
+    public void changeFsCopyButtonActionLink(String courseId, String feedbackSessionName, String newActionLink) {
+        String id = "button_fscopy" + "-" + courseId + "-" + feedbackSessionName;
+        By element = By.id(id);
+        waitForElementPresence(element);
+        
+        JavascriptExecutor js = (JavascriptExecutor) browser.driver;
+        js.executeScript("document.getElementById('"+id+"').setAttribute('data-actionlink', '"+newActionLink+"')");
+        
+    }
+    
     public void clickFsCopyButton(String courseId, String feedbackSessionName) {
         By element = By.id("button_fscopy" + "-" + courseId + "-" + feedbackSessionName);
         waitForElementPresence(element);
@@ -334,6 +344,10 @@ public class InstructorHomePage extends AppPage {
     
     public void waitForModalToLoad() {
         waitForElementPresence(By.id(Const.ParamsNames.COPIED_FEEDBACK_SESSION_NAME));
+    }
+    
+    public void waitForModalErrorToLoad() {
+        waitForElementPresence(By.id("fs-copy-modal-error"));
     }
     
     public void clickFsCopySubmitButton() {

